### ISSUE TITLE
Found the solution: Flutter 2.5 update causes app to freeze if the cursor has to scroll out of view.

### ISSUE THREAD

**Issue by Satsrag** [suragch/mongol#22]
The issue occurred in `paint(PaintingContext context, Offset offset)` method of `MongolRenderEditable` class .
The has issue code:

```dart
  @override
  void paint(PaintingContext context, Offset offset) {
    _layoutText(
        minHeight: constraints.minHeight, maxHeight: constraints.maxHeight);
    if (_hasVisualOverflow && clipBehavior != Clip.none) {
      _clipRectLayer = context.pushClipRect(
        needsCompositing,
        offset,
        Offset.zero & size,
        _paintContents,
        clipBehavior: clipBehavior,
        oldLayer: _clipRectLayer,
      );
    } else {
      _clipRectLayer = null;
      _paintContents(context, offset);
    }
    _paintHandleLayers(context, getEndpointsForSelection(selection!));
  }

  ClipRectLayer? _clipRectLayer;
```

The solution code:

```dart
  @override
  void paint(PaintingContext context, Offset offset) {
    _layoutText(
        minHeight: constraints.minHeight, maxHeight: constraints.maxHeight);
    if (_hasVisualOverflow && clipBehavior != Clip.none) {
      _clipRectLayer.layer = context.pushClipRect(
        needsCompositing,
        offset,
        Offset.zero & size,
        _paintContents,
        clipBehavior: clipBehavior,
        oldLayer: _clipRectLayer.layer,
      );
    } else {
      _clipRectLayer.layer = null;
      _paintContents(context, offset);
    }
    _paintHandleLayers(context, getEndpointsForSelection(selection!));
  }

  final LayerHandle<ClipRectLayer> _clipRectLayer = LayerHandle<ClipRectLayer>();
```

The `_clipRectLayer ` has been disposed when call `context.pushClipRect` method. So use `LayerHandle` to prevent `_clipRectLayer` from disposing. The solution code copy from flutter source `RenderEditable`.

**Comment by suragch** [comment]
That's great! I wish I had had your solution before! In the current version of the `mongol` package I've completely removed `MongolRenderEditable` and the other related classes from the library. My hope has been to reimplement them based on the [super_editor](https://github.com/superlistapp/super_editor) package, but I haven't had time. However, if you've found a solution to the previous error, it may be faster to use your solution first. What do you think about adding your solution on a new branch? Then we can test it. If everything is working we can merge it into the main branch.

**Comment by Satsrag** [comment]
I try to go back to the previous version of the `mongol` package to check out the new branch and then add the solution. But when I merge it to the master branch, some `MongolTextFiled` related code was deleted again. 
So I think checkout the new branch from the current version of the `mongol` package and then copy the `MongolTextFiled` related code from the previous version. Then add the solution and others?

**Comment by suragch** [comment]
Yes, let's not merge into the master branch for now. Let's create a new branch. Then people can use that branch if they need your fixed `MongolTextField`. Should I create a new branch?

**Comment by Satsrag** [comment]
Yes .

**Comment by suragch** [comment]
I created a new branch from master named `editor-fixes`. Feel free to modify this branch however you like.

**Comment by Satsrag** [comment]
I `pull request` to the `editor-fixes` branch #23. Please search `todo editor-fixes` to check my modified part.

**Comment by suragch** [comment]
I think the `editor-fixes` branch is almost ready to merge into `master`. It's really nice to see the editor widget working again. 

I've pushed a few minor cleanup changes to that branch. Before we merge, though, we should evaluate these issues:

- The items you mention in [this comment](https://github.com/suragch/mongol/pull/24#issue-1361079487). Let's create new issues to track each one. Are they serious enough that we need to solve them before merging?
- In my testing, I discovered that the Keyboard demo in the example project isn't working very well. At first it was crashing, but I added a fix for that. But it still isn't maintaining selection. And when you select text and then input a character, it gets inserted at the top. This is probably a problem with the keyboard demo logic but it might also be a problem with the editable widget. We should solve this before merging. (My testing was with the Flutter app running on macOS).

**Comment by suragch** [comment]
@Satsrag I've merged your changes into the master branch. Everything seems to be working well enough. I need to update the documentation and then I'll publish the update to Pub as Version 3.1.0. If you know of any issues that should be fixed first, please open another issue.

**Comment by suragch** [comment]
I've published your fixes in version 3.1. In addition to fixing the scrolling issue, it's also great to have the arrow keys working on web and desktop. Thank you very much!
